### PR TITLE
Added documentation for what happens when you start Galaxy

### PR DESCRIPTION
**Pull request by kxk302**
Documented call graph and what happens for each step of a function call, to explain what happens when you start Galaxy via run.sh script.

### CODE REVIEW

**Comment by dannon** [comment]
@kxk302 This is a great idea; do you plan on adding all the pre-python stuff from `common_startup.sh`, etc?  That'd be great to have.

**Comment by kxk302** [comment]
> Documented call graph and what happens for each step of a function call, to explain what happens when you start Galaxy via run.sh script.

Thanks @dannon! I will probably do the pre-python stuff as a separate task (I also plan to document all the steps involved in uploading a file to Galaxy).

**Comment by dannon** [comment]
@kxk302 that's great, thanks!

File: doc/source/dev/starting_galaxy.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  * Command.run()
+    * Setup default options, validate command line arguments, then call ServeCommand's command() method
+  * ServeCommand.command()
+    * Conditionally, stop the daemon 
+    * If a config file is required, make sure a config file is provided on the command line arguments.
+    * Possible subcommands: start, stop, restart, status (or None)
+    * Call loadserver() method in ./galaxy/lib/galaxy/util/pastescript/loadwsgi.py
+      * Calls loadobj()  -> calls LoaderContext's create() -> calls _Server's invoke()  
+    * Call loadapp() method in ./galaxy/lib/galaxy/util/pastescript/loadwsgi.py
+      * Calls loadobj()  -> calls LoaderContext's create() -> calls _App's invoke() -> calls fix_call() on app_factory (in ./galaxy/lib/galaxy/webapps/galaxy/buildapp.py)
```
Comment (mvdbeek) [inline doc/source/dev/starting_galaxy.md:31]: This part is specific to Paste. This is also not the typical way Galaxy starts if you start it via `run.sh` which uses uwsgi by default. If you can just add a note about that I think that'd be sufficient. Galaxy is a WSGI app, so in principle all python WSGI servers can run the Galaxy app (paster, gunicorn and uwsgi are known to work and have some level of support).

File: doc/source/dev/starting_galaxy.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          * Set GalaxyTagHandler (in ./galaxy/lib/galaxy/model/tags.py)
+            * Handles CRUD operations for tagging datasets, workflows, visualizations, etc
+          * Set DatasetCollectionManager (in ./galaxy/lib/galaxy/managers/collections.py) 
+          * Configure HistoryManager
+          * Configure HistoryDatasetAssociation (HDA) Manager
+          * Create WorkflowsManager
+            * Handles CRUD operations related to workflows
+          * Configure DependencyResolverView (a RESTful interface to galaxy.tool_util.deps.DependencyResolver)
+          * Configure TestDataResolver
+          * Configure LibraryManager (Data libraries under Shared Data on use galaxy.org)
```
Comment (mvdbeek) [inline doc/source/dev/starting_galaxy.md]:
```suggestion
          * Configure LibraryManager (Data libraries under Shared Data)
```

**Review by mvdbeek — COMMENTED** [review]
Nice, I think that'll be very useful!

File: doc/source/dev/starting_galaxy.md (hunk 1 above)
Comment (kxk302) [reply doc/source/dev/starting_galaxy.md:31]: I added the note and pushed.

File: doc/source/dev/starting_galaxy.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        * Create UniverseApplication  (in ./galaxy/lib/galaxy/webapps/galaxy/buildapp.py)
+          * Create GalaxyAppConfiguration instance from config file specified on the command line, then check for errors
+          * Configure logging (in ./galaxy/lib/galaxy/config/__init__.py)
+          * Create ExecutionTimerFactory
+            * Which has a GalaxyStatsdClient field
+          * Configure FluentTraceLogger
+          * Create the application stack (uWSGi, Paste, or Webless)
+            * App registers post fork functions to start application stack threads   
+          * Create GalaxyQueueWorker instance (a worker for Galaxy's queues, used for dispatching control tasks. Sqlite instance)
+          * Configure tool shed registery
```
Comment (jdavcs) [inline doc/source/dev/starting_galaxy.md]:
```suggestion
          * Configure tool shed registry
```

File: doc/source/dev/starting_galaxy.md (hunk 1 above)
Comment (mvdbeek) [reply doc/source/dev/starting_galaxy.md:31]: Now it sounds like the entire startup is completely different. These paths all converge on instantiating a UniverseApplication instance. See for instance https://github.com/galaxyproject/galaxy/blob/278a0333724e84209f317bc200c86ad5165f575a/scripts/galaxy-main#L105 which is probably the easiest way to build a Galaxy app that is not serving the WSGI life-cycle.

Comment (kxk302) [reply doc/source/dev/starting_galaxy.md:31]: Should I revise the doc in anyway?

Comment (kxk302) [reply doc/source/dev/starting_galaxy.md:31]: Or can it be merged as is?

**Comment by kxk302** [comment]
> Documented call graph and what happens for each step of a function call, to explain what happens when you start Galaxy via run.sh script.

Added a test to test_tools_upload.py, which generates a randomly populated file, uploads it to Galaxy, then downloads it, and finally deletes it. There is a small random delay between these steps. The goal is massively parallelize this test via Locust to load test iRODS in Galaxy.

**Comment by kxk302** [comment]
> > Documented call graph and what happens for each step of a function call, to explain what happens when you start Galaxy via run.sh script.
> 
> @mvdbeek Added a test to test_tools_upload.py, which generates a randomly populated file, uploads it to Galaxy, then downloads it, and finally deletes it. There is a small random delay between these steps. The goal is massively parallelize this test via Locust to load test iRODS in Galaxy.

This is the command to run the test:

GALAXY_TEST_EXTERNAL=http://127.0.0.1:8081 GALAXY_CONFIG_MASTER_API_KEY="abcdefg" sh run_tests.sh -api ./lib/galaxy_test/api/test_tools_upload.py::ToolsUploadTestCase::test_upload_download_delete_of_randomly_generated_file

**Comment by kxk302** [comment]
Created new PR that has only the documentation changes.

https://github.com/galaxyproject/galaxy/pull/9918

**Pull request closed without merging**